### ISSUE TITLE
Add transaction helper

### ISSUE THREAD

**Issue by hadley** [rstats-db/DBI#104]
Something like this:

```R
withTransaction <- function(con, fun, ...) {
  dbBegin(con)
  tryCatch(
    {
      out <- fun(con)
      dbCommit(con)
    },
    error = function(e){
      dbRollback(e)
      stop(e)
    }
  )

  out
}
```

Ensures that rollback happens automatically on error - otherwise this is quite a lot of boilerplate to get correct

(I don't think this needs to be a generic, but I'm not 100% sure)

**Comment by hannesmuehleisen** [comment]
+1

**Comment by krlmlr** [comment]
How about:

```
withTransaction <- function(con, code) {
  dbBegin(con)
  on.exit(dbRollback(con), add = FALSE)
  out <- code
  dbCommit(con)
  on.exit(NULL, add = FALSE)
  out
})
```

If the behavior of dbBegin() and dbRollback() is specified and tested (https://github.com/rstats-db/DBItest/issues/36), this probably doesn't need to be a generic.

**Comment by hadley** [comment]
That look ok @jcheng5?

**Comment by jcheng5** [comment]
That's clever and it looks good to me. For the connection pool we will need this to be generic though. Our method will need to acquire and release a connection from the pool.

**Comment by krlmlr** [comment]
Connection pool? I'm intrigued ;-)

It remains unclear to me how you intend to implement withTransaction() for that connection pool. Wouldn't the code need to access the new connection? Isn't the acquisition and release of a new connection is a separate operation?

**Comment by hadley** [comment]
Right, it does need to be genernic.  The connection pool implementation method would look something like:

```R
withTransaction.pool <- function(con, code) {
  con <- dbConnect(con)
  on.exit(dbDisconnect(con))
  withTransaction(con, code)
```

(but in S4 obviously, I'm not sure if those are the verbs the pool will use. More on the pool soon)
  


```

**Comment by hadley** [comment]
And probably should be `dbWithTransaction()` to be consistent with other DBI generics.

**Comment by krlmlr** [comment]
In withTransaction.pool(), how can the code access the new connection? How about

```
dbWithNewConnection <- function(con, code) {
  ...
}
```

which doesn't solve the problem but allows executing both non-transacted and transacted code in a new connection?

**Comment by krlmlr** [comment]
Fixed in #110.
